File: src/apisniff/adapters/har.py
from __future__ import annotations

import base64
import json
from datetime import datetime
from urllib.parse import urlparse

from apisniff.models import CapturedFlow


def _parse_headers(header_list: list[dict]) -> dict[str, str]:
    """Group HAR headers by lowercased name.

    Per RFC 9110 multiple values for the same header are joined with ", ",
    except Set-Cookie which must remain newline-separated so each directive
    is individually parseable.
    """
    grouped: dict[str, list[str]] = {}
    for h in header_list:
        key = h["name"].lower()
        grouped.setdefault(key, []).append(h["value"])

    result: dict[str, str] = {}
    for key, values in grouped.items():
        if key == "set-cookie":
            result[key] = "\n".join(values)
        else:
            result[key] = ", ".join(values)
    return result


def _parse_timestamp(entry: dict) -> float:
    raw = entry.get("startedDateTime")
    if not raw:
        return 0.0
    try:
        normalized = raw.rstrip("Z")
        if raw.endswith("Z"):
            normalized += "+00:00"
        return datetime.fromisoformat(normalized).timestamp()
    except (ValueError, AttributeError):
        return 0.0


def har_to_flows(har_text: str) -> list[CapturedFlow]:
    data = json.loads(har_text)
    entries = data.get("log", {}).get("entries", [])
    flows: list[CapturedFlow] = []

    for entry in entries:
        req = entry.get("request", {})
        resp = entry.get("response", {})

        url = req.get("url", "")
        parsed = urlparse(url)

        path = parsed.path or "/"
        if parsed.query:
            path = path + "?" + parsed.query

        req_headers = _parse_headers(req.get("headers", []))
        resp_headers = _parse_headers(resp.get("headers", []))

        req_body = (req.get("postData", {}) or {}).get("text", "") or ""

        content = resp.get("content", {}) or {}
        resp_body_text = content.get("text", "") or ""
        if content.get("encoding") == "base64" and resp_body_text:
            resp_body: bytes = base64.b64decode(resp_body_text)
        else:
            resp_body = (
                resp_body_text.encode("utf-8")
                if isinstance(resp_body_text, str) else resp_body_text
            )

        flows.append(CapturedFlow(
            method=req.get("method", "GET"),
            host=parsed.hostname or "",
            path=path,
            url=url,
            request_headers=req_headers,
            request_body=req_body.encode("utf-8") if isinstance(req_body, str) else req_body,
            response_status=resp.get("status", 0),
            response_headers=resp_headers,
            response_body=resp_body,
            timestamp=_parse_timestamp(entry),
        ))

    return flows

File: internal/adapter/har.go
package adapter

import (
	"encoding/base64"
	"encoding/json"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/4LAU/apisniff/internal/model"
)

type harFile struct {
	Log struct {
		Entries []harEntry `json:"entries"`
	} `json:"log"`
}

type harEntry struct {
	StartedDateTime string      `json:"startedDateTime"`
	Request         harRequest  `json:"request"`
	Response        harResponse `json:"response"`
}

type harRequest struct {
	Method   string      `json:"method"`
	URL      string      `json:"url"`
	Headers  []harHeader `json:"headers"`
	PostData *struct {
		Text string `json:"text"`
	} `json:"postData"`
}

type harResponse struct {
	Status  int         `json:"status"`
	Headers []harHeader `json:"headers"`
	Content struct {
		Text     string `json:"text"`
		Encoding string `json:"encoding"`
	} `json:"content"`
}

type harHeader struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

func LoadHAR(path string) ([]model.CapturedFlow, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var har harFile
	if err := json.Unmarshal(data, &har); err != nil {
		return nil, err
	}
	flows := make([]model.CapturedFlow, 0, len(har.Log.Entries))
	for _, entry := range har.Log.Entries {
		parsed, err := url.Parse(entry.Request.URL)
		if err != nil || !isAbsoluteRequestURL(parsed) {
			continue
		}
		path := parsed.EscapedPath()
		if path == "" {
			path = "/"
		}
		if parsed.RawQuery != "" {
			path += "?" + parsed.RawQuery
		}
		reqBody := []byte{}
		if entry.Request.PostData != nil {
			reqBody = []byte(entry.Request.PostData.Text)
		}
		respBody := []byte(entry.Response.Content.Text)
		tags := []string{}
		if strings.EqualFold(entry.Response.Content.Encoding, "base64") && entry.Response.Content.Text != "" {
			if decoded, err := base64.StdEncoding.DecodeString(entry.Response.Content.Text); err == nil {
				respBody = decoded
			} else {
				// Keeping the undecoded base64 text would corrupt downstream
				// consumers; drop the body and mark the flow instead.
				respBody = nil
				tags = append(tags, "response_body_decode_error")
			}
		}
		flows = append(flows, model.CapturedFlow{
			Method:          model.FirstNonEmpty(entry.Request.Method, "GET"),
			Host:            parsed.Hostname(),
			Path:            path,
			URL:             entry.Request.URL,
			RequestHeaders:  parseHARHeaders(entry.Request.Headers),
			RequestBody:     reqBody,
			ResponseStatus:  entry.Response.Status,
			ResponseHeaders: parseHARHeaders(entry.Response.Headers),
			ResponseBody:    respBody,
			BodyEncoding:    "base64",
			Tags:            tags,
			Timestamp:       parseHARTimestamp(entry.StartedDateTime),
		})
	}
	return flows, nil
}

func parseHARHeaders(headers []harHeader) map[string]string {
	grouped := map[string][]string{}
	for _, header := range headers {
		if header.Name == "" {
			continue
		}
		key := strings.ToLower(header.Name)
		grouped[key] = append(grouped[key], header.Value)
	}
	return joinHeaderValues(grouped)
}

func parseHARTimestamp(raw string) float64 {
	if raw == "" {
		return 0
	}
	t, err := time.Parse(time.RFC3339Nano, raw)
	if err != nil {
		return 0
	}
	return float64(t.UnixNano()) / 1e9
}

func isAbsoluteRequestURL(parsed *url.URL) bool {
	return parsed != nil && parsed.Scheme != "" && parsed.Host != ""
}
